src/coreclr/src/jit/lsra.cpp

```diff
@@ -2994,291 +2989,584 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
         }
+    }
 
-        if ((candidateBit & callerCalleePrefs) != RBM_NONE)
-        {
-            score |= CALLER_CALLEE;
-        }
+    // Apply the THIS_ASSIGNED heuristic. Only applies if we have freeCandidates.
+    if (!found && (prevRegRec != nullptr) && (freeCandidates != RBM_NONE))
+    {
+        found = selector.applySelection(THIS_ASSIGNED, freeCandidates & preferences & prevRegBit);
+    }
 
-        // The register is considered unassigned if it has no assignedInterval, OR
-        // if its next reference is beyond the range of this interval.
-        if (!isAssigned(physRegRecord, lastLocation ARM_ARG(currentInterval->registerType)))
+    // Compute the sets for COVERS, OWN_PREFERENCE, COVERS_RELATED, COVERS_FULL and UNASSIGNED together,
+    // as they all require similar computation.
+    regMaskTP coversSet        = RBM_NONE;
+    regMaskTP preferenceSet    = RBM_NONE;
+    regMaskTP coversRelatedSet = RBM_NONE;
+    regMaskTP coversFullSet    = RBM_NONE;
+    if (freeCandidates != RBM_NONE)
+    {
+        preferenceSet              = (selector.candidates & preferences);
+        regMaskTP coversCandidates = (preferenceSet == RBM_NONE) ? selector.candidates : preferenceSet;
+        for (; coversCandidates != RBM_NONE;)
         {
-            score |= UNASSIGNED;
-        }
+            regMaskTP coversCandidateBit = genFindLowestBit(coversCandidates);
+            coversCandidates &= ~coversCandidateBit;
+            regNumber coversCandidateRegNum = genRegNumFromMask(coversCandidateBit);
 
-        bool foundBetterCandidate = false;
-
-        if (score > bestScore)
-        {
-            foundBetterCandidate = true;
-        }
-        else if (score == bestScore)
-        {
-            // Prefer a register that covers the range.
-            if (bestLocation <= lastLocation)
+            // If we have a single candidate we don't need to compute the preference-related sets, but we
+            // do need to compute the unassignedSet.
+            if (!found)
             {
-                if (nextPhysRefLocation > bestLocation)
+                // Find the next RefPosition of the register.
+                LsraLocation nextIntervalLocation    = getNextIntervalRef(coversCandidateRegNum, regType);
+                LsraLocation nextPhysRefLocation     = getNextFixedRef(coversCandidateRegNum, regType);
+                LsraLocation coversCandidateLocation = Min(nextPhysRefLocation, nextIntervalLocation);
+#ifdef TARGET_ARM
+                if (currentInterval->registerType == TYP_DOUBLE)
                 {
-                    foundBetterCandidate = true;
+                    LsraLocation otherNextPhysRefLocation =
+                        Min(nextFixedRef[coversCandidateRegNum + 1], nextIntervalRef[coversCandidateRegNum + 1]);
+                    coversCandidateLocation = Min(coversCandidateLocation, otherNextPhysRefLocation);
```

Comment:
Actually, this code should have been eliminated, since the methods above now take the type into account for ARM. Also, it appears that the spill code was actually taking the max, which doesn't make sense. So, on further thought it seems best just to do the logical thing and take the Min, and I'll take another look at the diffs.